internal/pkg/client/patch/patch.go

```diff
@@ -0,0 +1,210 @@
+	if err != nil {
+		return patch, fmt.Errorf("could not serialize intended configuration from %v", desired)
+	}
+	gvk := desired.GetObjectKind().GroupVersionKind()
+
+	versionedObject, err := Scheme.New(gvk)
+	_, unversioned := Scheme.IsUnversioned(desired)
+	switch {
+	case runtime.IsNotRegisteredError(err) || unversioned:
+		preconditions := []mergepatch.PreconditionFunc{
+			mergepatch.RequireKeyUnchanged("apiVersion"),
+			mergepatch.RequireKeyUnchanged("kind"),
+			mergepatch.RequireMetadataKeyUnchanged("name"),
+		}
+		patch.Data, err = jsonmergepatch.CreateThreeWayJSONMergePatch(last, modified, currentb, preconditions...)
+		patch.Type = types.MergePatchType
+		if err != nil {
+			name, err := metadataAccessor.Name(desired)
+			if mergepatch.IsPreconditionFailed(err) {
+				return patch, fmt.Errorf("at least one of apiVersion, kind and name was changed for %s/%s", gvk, name)
+			}
+			return patch, err
+		}
+	case err != nil:
+		return patch, err
+	case err == nil:
+		// Compute a three way strategic merge patch to send to server.
+		patch.Type = types.StrategicMergePatchType
+		lookupPatchMeta, err := strategicpatch.NewPatchMetaFromStruct(versionedObject)
+		if err != nil {
+			return patch, err
+		}
+		patch.Data, err = strategicpatch.CreateThreeWayMergePatch(last, modified, currentb, lookupPatchMeta, true)
+		if err != nil {
+			return patch, err
+		}
+	}
+	return patch, nil
+}
+
+// GetMergePatch - generate merge patch from original and modified objects
+func GetMergePatch(original, modified runtime.Object) (Patch, error) {
+	originalJSON, err := json.Marshal(original)
+	if err != nil {
+		return Patch{}, err
+	}
+
+	modifiedJSON, err := json.Marshal(modified)
+	if err != nil {
+		return Patch{}, err
+	}
+
+	patch, err := jsonpatch.CreateMergePatch(originalJSON, modifiedJSON)
+	if err != nil {
+		return Patch{}, err
+	}
+	return Patch{types.MergePatchType, patch}, nil
+}
+
+// GetLastApplied retrieves the original configuration of the object
+// from the annotation, or nil if no annotation was found.
+func GetLastApplied(obj runtime.Object) ([]byte, error) {
+	annots, err := metadataAccessor.Annotations(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	if annots == nil {
+		return nil, nil
+	}
+
+	original, ok := annots[v1.LastAppliedConfigAnnotation]
+	if !ok {
+		return nil, nil
+	}
+
+	return []byte(original), nil
+}
+
+// SetLastApplied patch
+func SetLastApplied(obj runtime.Object) error {
+	modified, err := GetModifiedConfiguration(obj, false)
+	if err != nil {
+		return err
+	}
+	annots, err := metadataAccessor.Annotations(obj)
+	if err != nil {
+		return err
+	}
+
+	if annots == nil {
+		annots = map[string]string{}
+	}
+
+	annots[v1.LastAppliedConfigAnnotation] = string(modified)
+	return metadataAccessor.SetAnnotations(obj, annots)
+}
+
+// GetModifiedConfiguration retrieves the modified configuration of the object.
+// If annotate is true, it embeds the result as an annotation in the modified
```

Comment:
what is `the result`?  `The returned object will contain a serialized annotation of itself`?

Reply:
renamed the function and comments. Not splitting for now.